### PR TITLE
[lodash] add type guard for `isObjectLike`

### PR DESCRIPTION
**Pull request by otomad**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `pnpm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:

- [x] Provide a URL to documentation or source code which provides context for the suggested changes:
    - https://lodash.com/docs/4.17.15#isObjectLike
    - https://lodash.com/docs/4.17.15#isPlainObject
- [x] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the `package.json`.

---

Fix #68867

- `isObject`: After testing, I found that the `object` type itself contains function type, so the current type declaration is correct and does not need to be modified to `value is object | (...any[]) => any`. In addition, I added some additional tests.
- `isObjectLike`: As can the reason above, the `object` type includes functions, so directly using `value is object` will cause an error. For example, when `value` is a function, `if(!isObjectLike(value))` will cause `value` to become a `never` type. So I modified it to determine whether value is an object type, but not a function type.
- ~~`isPlainObject`: In the discussion, it was mentioned that if `isPlainObject` is type guard, the inverse type guard would be inaccurate. However, I hacked by adding a unique Symbol that will never be used in the `isPlainObject` type. In this way, although an additional Symbol property appears in `if(isPlainObject(value))`, it will never be used and will not be affected; while `if(!isPlainObject(value))` will not cause any change to the type so keep the original type of `value`.~~

Reference:

https://github.com/DefinitelyTyped/DefinitelyTyped/discussions/68867#discussioncomment-8671502

### CODE REVIEW

File: types/lodash/lodash-tests.ts [hunk 1]
```diff
@@ -4542,6 +4558,17 @@ fp.now(); // $ExpectType number
 
 // _.isPlainObject
 {
+    const value: Function | (() => void) | { foo: string; } | string = anything;
+
+    if (_.isPlainObject(value)) {
+        const result: Function | (() => void) | { foo: string; } = value;
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:4564]: This doesn't look right. The only "plain object" should be `{foo: string}`. Can you use $ExpectType so it's easier to see the exact types in each block?

Also, you should add tests for arrays and `new Foo()` as shown in the lodash documentation.

File: types/lodash/isPlainObject-symbol.d.ts [hunk 2]
```diff
@@ -0,0 +1,3 @@
+// @ts-expect-error
```
Comment (aj-r) [inline types/lodash/isPlainObject-symbol.d.ts:1]: It's a little weird to include this in production code. Is this necessary? I'd like to find a different approach if possible.

File: types/lodash/lodash-tests.ts [hunk 3]
```diff
@@ -4534,6 +4542,14 @@ fp.now(); // $ExpectType number
 
 // _.isObjectLike
 {
+    const value: Function | (() => void) | { foo: string; } | string = anything;
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:4545]: Add tests for array, null, and undefined.

**Review by aj-r — CHANGES_REQUESTED** [review]

File: types/lodash/lodash-tests.ts (hunk 3 above)
Comment (aj-r) [reply types/lodash/lodash-tests.ts:4545]: Maybe `new Foo()` too

Comment (otomad) [reply types/lodash/lodash-tests.ts]: Added.

File: types/lodash/lodash-tests.ts (hunk 1 above)
Comment (otomad) [reply types/lodash/lodash-tests.ts]: Added undefined, null, array, and Foo class, and also modified the type declaration so that the result excludes functions and arrays and only includes `{foo: string}` and `new Foo()`.

Why include `new Foo()` which is not expected? This is because TypeScript can only recognize ["duck typing"](https://en.wikipedia.org/wiki/Duck_typing). It does not check the **prototype** of the object, but checks whether the individual properties of the object match. Therefore, TypeScript does not think that there is any difference between an instance of the `Foo` class and a plain object containing all the properties of `Foo`.

Also, at least it ensures that the inverse type guard (`if (!_.isPlainObject(value))`) is not affected.

File: types/lodash/isPlainObject-symbol.d.ts (hunk 2 above)
Comment (otomad) [reply types/lodash/isPlainObject-symbol.d.ts]: I don't understand whether you mean the `@ts-expect-error` comment, the `IS_PLAIN_OBJECT` Symbol, or this .d.ts file is unnecessary?

For the `@ts-expect-error` comment, I've removed it in the lastest commit.

For the `IS_PLAIN_OBJECT` Symbol, this is because TypeScript needs to narrow a type by typing like `A & B` in the type guard so that its inverse type guard is not affected. In this case, A is the `object` and B needs to be a type that is unlikely to exist. I used Symbol here because Symbol is unique and users can hardly override it. In addition, I also planned to use `{ __proto__: Object | null }`, which is more appropriate to the actual function of `isPlainObject`. However, if the users also define the `__proto__` property in their value type, then the result will be incorrect.

For the .d.ts file, In fact, I can define this Symbol directly in the common/lang.d.ts file. Because of the fp.d.ts file, which generated by generate-fp will not find this Symbol, so import the Symbol from external instead.

File: types/lodash/lodash-tests.ts [hunk 4]
```diff
@@ -4542,6 +4568,19 @@ fp.now(); // $ExpectType number
 
 // _.isPlainObject
 {
+    const value: Function | (() => void) | { foo: string; } | string | undefined | null | string[] | Foo = anything;
+
+    if (_.isPlainObject(value)) {
+        value; // $ExpectType ({ foo: string; } & object & { [IS_PLAIN_OBJECT]: undefined; }) | (Foo & object & { [IS_PLAIN_OBJECT]: undefined; })
+        const result: { foo: string; } | Foo = value;
```
Comment (aj-r) [inline types/lodash/lodash-tests.ts:4575]: Is `|Foo` necessary here? This seems to indicate that the `IS_PLAIN_OBJECT` trick isn't actually working. Maybe it's better to provide no type guard rather than provide a false sense of security. Anyone who wants the type guard and doesn't care to differentiate plain objects vs. classes should use `isObject` or `isObjectLike` instead.

File: types/lodash/isPlainObject-symbol.d.ts (hunk 2 above)
Comment (aj-r) [reply types/lodash/isPlainObject-symbol.d.ts]: Yeah I meant `@ts-expect-error`.

`generate-fp` doesn't really work anymore. We should probably just delete it. You'll need to update `fp.d.ts` manually.

**Review by aj-r — CHANGES_REQUESTED** [review]

File: types/lodash/lodash-tests.ts (hunk 4 above)
Comment (otomad) [reply types/lodash/lodash-tests.ts:4575]: Fine, I will remove the type guard for `isPlainObject`.

**Comment by otomad** [comment]
There is a known unexpected behavior for `isObjectLike`. When the type of value is `(() => void) | {}`, `isObjectLike` will narrow the type of value to `(() => void) | {}` and the inverse type guard to `never`.

I submit an issue https://github.com/microsoft/TypeScript/issues/61597, and it's not clear if this is an expected behavior and can be fixed.

**Comment by aj-r** [comment]
> There is a known unexpected behavior for `isObjectLike`. When the type of value is `(() => void) | {}`, `isObjectLike` will narrow the type of value to `(() => void) | {}` and the inverse type guard to `never`.
> 
> I submit an issue [microsoft/TypeScript#61597](https://github.com/microsoft/TypeScript/issues/61597), and it's not clear if this is an expected behavior and can be fixed.

Weird... maybe TS treats `{}` as NOT extending `object` since it could also represent a `string`, `number`, etc. Although if that were the case, I would expect the behavior to be the exact opposite of what you describe...

That raises an interesting question though: how does this type guard behave for a plain `{}` type? Ideally I think both the positive and inverse type guards should not narrow at all, and leave the type as `{}`. Or maybe narrow `object` in the positive type guard and `string | number` in the negative. It might be worth adding a test since this is a weird case.

**Comment by otomad** [comment]
According to the reply https://github.com/microsoft/TypeScript/issues/61597#issuecomment-2817508079, it seems that there is no perfect solution currently, we need negated types for the use case to be achievable in a straightforward way, or do something like [`object & {call?: never}`](https://stackoverflow.com/questions/52692606/how-to-declare-a-type-in-typescript-that-only-includes-objects-and-not-functions) which isn't perfect either.

Perhaps we can only continue to complete it after the negated types are supported in TypeScript.
